### DISCUSSION TITLE
Generic relationship

### DISCUSSION THREAD

**Discussion by klaydev** [sqlalchemy/sqlalchemy#10518] [Usage Questions] [answered]
Hi. I want to make a system of promo codes for orders. The question arose, how can I store these promo codes in an order if the promo codes are different and are in different tables? Here is an example:

```py
class Order(Base):
    __tablename__ = "order"
    
    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    promocode_id: Mapped[int] = mapped_column(ForeignKey(???))
    promocode: Mapped[???] = relathionship(???)
    
 class GlobalPromocode(Base):
     __tablename__ = "global_promocodes"
     
     id: Mapped[int] = mapped_column(primary_key=True, index=True)
     name: Mapped[str] = mapped_column(nullable=False)
     value: Mapped[float] = mapped_column(nullable=False)
     
 class CategoryPromocode(Base):
     __tablename__ = "category_promocodes"
     
     id: Mapped[int] = mapped_column(primary_key=True, index=True)
     name: Mapped[str] = mapped_column(nullable=False)
     value: Mapped[float] = mapped_column(nullable=False)
```
I want to implement this behavior:
```py
order = Order()
promocode = GlobalPromocode()
order.promocode = promocode
```
And
```py
order = Order()
promocode = CategoryPromocode()
order.promocode = promocode
```
Thank you in advance for your time and help

**Comment by CaselIT** [comment]
Hi,

You can look into inheritance, so that your two promotions inherit from a common promotion class.
More info in this page of the docs: https://docs.sqlalchemy.org/en/20/orm/inheritance.html

**Reply by klaydev** [reply]
Thank you. I tried, but now I have this error: `sqlalchemy.exc.InvalidRequestError: Can't copy ForeignKey object which refers to non-table bound Column Column(None, NullType(), table=None, primary_key=True, nullable=False, default=CallableColumnDefault(<function uuid4 at 0x000001F5C9BFB600>))`. The fact is that I have a Foreign Key, in which you need to specify a table, I pass it to him ClassName.id , instead of the string "table_name.id ", because this Foreign Key can be in several tables

**Reply by CaselIT** [reply]
You can't have a fk that points to several tables. To make this work you probably need to use single table inheritance or the joined table one

**Reply by zzzeek** [reply]
not a real foreign key.

**Comment by zzzeek** [comment, marked as answer]
We have complete examples for four different varieties of "generic foreign key" including the non-normalized one that people often want to use, these are here: https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.generic_associations
